### ISSUE TITLE
TLS/SSL Secure Socket

### ISSUE THREAD

**Issue by galvesribeiro** [rdavisau/sockets-for-pcl#4]
Greate work! I was looking for ages to avoid the interface abstractions and instead a PCL library for sockets. Thank you!

So, any changes on short term to have a TLS/SSL Secure Socket implementation supported?

Need any help? We can discuss and maybe I can help with code.

Thanks!

**Comment by galvesribeiro** [comment]
As far as I can check on the code, you are lacking the implementation of SslStream on it... If you need any help please let me know and I'm glad to make a PR.

Thanks

**Comment by rdavisau** [comment]
Yes, I had also been looking for a way to avoid working with interface abstractions for a while, so finally bit the bullet and put together this library.

TLS/SSL is not my area of expertise, but if you want to give your thoughts on the changes needed or put together a PR that would be awesome! Let me know what you think and if you need anything from me. 

Ryan

**Comment by galvesribeiro** [comment]
OK, I'll have a look over the weekend. I don't understood correctly the way you made several project behind a single PCL and then "proxed" then with a dumb PCL but I know it is working and well. My unique point now is understood that, and then I will have a look on how to implement the SslStream. Looks like you not actually implementing anything in the PCL, they are just proxies. If I'm correct, we should be able to easily proxy the other missing classes since there is a implementation for each one of them in the specific platform... If you can put together how you designed the solution, that would help me a lot.

Thank you.
Gutemberg

**Comment by rdavisau** [comment]
Sure -

The technique being used here is often referred to as the "bait and switch pcl method". If you google this you will find a few blog posts that explain the topic. I can explain a bit more later but the blog posts should set you in the right direction. In terms of what the implications are:

- any class or method you want the end user to be able to code against must be defined in the Sockets.Plugin project

- every platform project must implement the same classes and methods, but can include anything extra they need to implement the required functionality

- in sockets-for-pcl, you usually don't need to implement code in the platform projects, only "Sockets.Implementation.Net" and "Sockets.Implementation.WinRT", as this is an abstraction over the .NET and WinRT frameworks and the code is the same between all .NET platforms (iOSx2, Android, Windows Desktop), and all WinRT platforms (WinPhonex2, Windows Store). The files from the implementation projects are linked in to the actual platform projects. 

- because of the above, if you create any new files in the Sockets.Implementation projects you must link them in to their respective platform projects using Visual Studio. 

To summarise that: 
- Add new classes and methods to Sockets.Plugin
- Implement the new methods against .NET in Sockets.Implementation.Net and against WinRT in Sockets.Implementation.WinRT
- Link (not copy) any new files into each platform's project

Let me know if you have any questions!

**Comment by galvesribeiro** [comment]
Thank you! I'm cloning right away and checking what I can do, will get in touch. Thanks!

**Comment by galvesribeiro** [comment]
Hello Ryan,

Thank you for the explanation and for the great design of the solution. Now I got the idea that "There is no PCL at all!" :)

So, I've created a PR with the implementation of SSL/TLS on your project. It was kept basic just supporting server certificate that is 90% of the cases. 

Its optional for current users of your NuGet package and will not break any users since I'm using a simple bool optional parameter on ConnectAsync that has its default value to false, so no one will be affected.

I really appreciate if you can merge this PR and make it available on a new release on NuGet package so I can use it on my project.

Thanks!

Let me know if you have any problems.
Gutemberg

**Comment by rdavisau** [comment]
Thanks, this looks good! I'm pleased that the change is non-breaking for existing users. 

Can you explain a little more about the support for server certificate vs client certificate and what that means? My understanding is that it is validating server certificates but does not support situations which require client to provide a certificate which as you say is very rare. Is that correct? I'd also like to know what other kinds of validations would come under the `TODO` comment you have. 

I have a few specific comments which I'll add now in line with the PR - once you've confirmed these I'll merge and publish a prerelease package you can use to make sure everything works. 

Great stuff and very quick!

**Comment by rdavisau** [comment]
Ok, I have added my comments! 

I am not very familiar with `SecureStream`  so some of my assumptions may be wrong, but I think the changes are all very small. Because I haven't brought in any logging dependency, I'd prefer to keep any `Console.WriteLine's` out but we can leave them in for prerelease if it will help with your testing. Let me know what you think and once you've gone through I'll do the package up for you. 

Ryan

**Comment by galvesribeiro** [comment]
Sure, no problem.

So, about the certificates... Imagine that in the HTTP scenario, only the server has the certificate, right? And if the user don't trust the issuer, a warning alert is presented by the browser. The same happens on RAW Sockets.

The client certificate requires that the client has a certificate that the server trust, along with the server certificate itself. It is a pain to install a certificate in a mobile device. It is usually used for old technologies that doesn't have a CA chain easily updatable.

I'm a bit noob on Git so, what should I do now? Make the changes here and add a new PR?

Thanks

**Comment by rdavisau** [comment]
Ok, so it's similar to HTTP like you say. Cool. I'll merge and pack now :tada:

**Comment by rdavisau** [comment]
Ok, I've added a prerelease version of sockets-for-pcl to nuget with the secure support added: [1.1.4-pre](https://www.nuget.org/packages/rda.SocketsForPCL/1.1.4-pre). I had to add a small change in to account for the fact that WP8 has a different SocketProtectionLevel enum, it should be the correct value.

Please enable installation of prerelease packages and give it a try. Thanks for your help!

**Comment by galvesribeiro** [comment]
Thanks Ryan,

I’m testing right now. Will get in touch with you soon with a feedback.

Thanks

**Comment by galvesribeiro** [comment]
Hello Ryan,

Completely tested. Works like a charm. You can promote it now to release.

Thank you!

**Comment by rdavisau** [comment]
Hey Gutemberg, 

Thanks so much for your work developing and testing the TLS functionality! 
I've pushed the prelease package out as 1.1.4 and it's available on NuGet now. 

Thanks again for your help. 

Ryan

**Comment by galvesribeiro** [comment]
No problem, you can reach me on skype at the same alias as here. As soon as I have time, I'll get back to check the client certificate and add the support here.
Thanks

**Comment by rdavisau** [comment]
Great! I'll look forward to it :tada:
